components/annotation-services/ncr/src/main/java/org/phenotips/services/annotations/ncr/internal/NCRServlet.java

```diff
@@ -0,0 +1,184 @@
+ *
+ * @version $Id$
+ * @since 1.4
+ */
+public class NCRServlet extends HttpServlet
+{
+    /** The logging object. */
+    private static final Logger LOGGER = LoggerFactory.getLogger(NCRServlet.class);
+
+    /** The jackson object mapper. */
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    /** The Neural Concept Recognizer service url. */
+    private static final String SERVICE_URL = "https://ncr.ccm.sickkids.ca/curr/annotate/";
+
+    private static final String TEXT_LABEL = "text";
+
+    private static final String CONTENT_LABEL = "content";
+
+    private static final String APPLICATION_JSON = "application/json";
+
+    private static final long serialVersionUID = -8185410465881809448L;
+
+    /**
+     * Gets the service URL for the Neural Concept Recognizer.
+     *
+     * @return the service URL as string
+     */
+    public String getServiceUrl()
+    {
+        return SERVICE_URL;
+    }
+
+    /**
+     * Translates and forwards the post request from the clinical-text-analysis-extension. Writes the response expected
+     * by the clinical-text-analysis-extension.
+     *
+     * @param request the {@link HttpServletRequest request} object
+     * @param response the {@link HttpServletResponse response} object
+     */
+    @Override
+    protected void doPost(final HttpServletRequest request, final HttpServletResponse response)
+    {
+        try {
+            // Translate and forward the request to the NCR service.
+            final String text = forwardPost(request);
+            // Set content type for response.
+            response.setContentType(APPLICATION_JSON);
+            // Get response writer.
+            final PrintWriter writer = response.getWriter();
+            // Forward the retrieved and translated response.
+            writer.append(text);
+        } catch (final URISyntaxException e) {
+            LOGGER.error("Invalid URI syntax: {}", e.getMessage());
+            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+        } catch (final MalformedURLException e) {
+            LOGGER.error("The NCR service url [{}] is invalid: {}", SERVICE_URL, e.getMessage());
+            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+        } catch (final IOException e) {
+            LOGGER.error("Error with the request for or response from [{}]: {}", SERVICE_URL, e.getMessage());
+            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+        } catch (final Exception e) {
+            LOGGER.error("An unexpected error occurred: {}", e.getMessage());
+            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    /**
+     * Translates the request parameters to the format expected by the NCR service located at {@link #SERVICE_URL}.
+     *
+     * @param request the incoming request
+     * @return a list of name-value pairs, in format expected by the NCR service
```

Comment:
To make it more clear both what the code does and which version of NCR this is supposed to work with I would add an example (or specification) of the expected format in the comments. Same with reply format

Reply:
Fixed